ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoLocalAssembler-impl.h

```diff
@@ -0,0 +1,232 @@
+
+    NodalMatrixType mass_operator;
+    mass_operator.setZero(ShapeFunction::NPOINTS, ShapeFunction::NPOINTS);
+
+    auto Mgp =
+        local_M.template block<nonwet_pressure_size, nonwet_pressure_size>(
+            nonwet_pressure_matrix_index, nonwet_pressure_matrix_index);
+    auto Mgx = local_M.template block<nonwet_pressure_size, cap_pressure_size>(
+        nonwet_pressure_matrix_index, cap_pressure_matrix_index);
+
+    auto Mlp = local_M.template block<cap_pressure_size, nonwet_pressure_size>(
+        cap_pressure_matrix_index, nonwet_pressure_matrix_index);
+
+    auto Mlx = local_M.template block<cap_pressure_size, cap_pressure_size>(
+        cap_pressure_matrix_index, cap_pressure_matrix_index);
+
+    NodalMatrixType laplace_operator;
+    laplace_operator.setZero(ShapeFunction::NPOINTS, ShapeFunction::NPOINTS);
+
+    auto Kgp =
+        local_K.template block<nonwet_pressure_size, nonwet_pressure_size>(
+            nonwet_pressure_matrix_index, nonwet_pressure_matrix_index);
+
+    auto Kgx = local_K.template block<nonwet_pressure_size, cap_pressure_size>(
+        nonwet_pressure_matrix_index, cap_pressure_matrix_index);
+
+    auto Klp = local_K.template block<cap_pressure_size, nonwet_pressure_size>(
+        cap_pressure_matrix_index, nonwet_pressure_matrix_index);
+
+    auto Klx = local_K.template block<cap_pressure_size, cap_pressure_size>(
+        cap_pressure_matrix_index, cap_pressure_matrix_index);
+
+    auto Bg = local_b.template segment<nonwet_pressure_size>(
+        nonwet_pressure_matrix_index);
+
+    auto Bl =
+        local_b.template segment<cap_pressure_size>(cap_pressure_matrix_index);
+
+    unsigned const n_integration_points =
+        _integration_method.getNumberOfPoints();
+
+    SpatialPosition pos;
+    pos.setElementID(_element.getID());
+    _process_data._material->setMaterialID(pos);
+
+    const Eigen::MatrixXd& perm = _process_data._material->getPermeability(
+        t, pos, _element.getDimension());
+    assert(perm.rows() == GlobalDim || perm.rows() == 1);
+    GlobalDimMatrixType permeability =
+        GlobalDimMatrixType::Zero(GlobalDim, GlobalDim);
+    if (perm.rows() == GlobalDim)
+        permeability = perm;
+    else if (perm.rows() == 1)
+        permeability.diagonal().setConstant(perm(0, 0));
+    // Note: currently only isothermal case is considered, so the temperature is
+    // assumed to be const
+    // the variation of temperature will be taken into account in future
+    for (unsigned ip = 0; ip < n_integration_points; ip++)
+    {
+        auto const& sm = _shape_matrices[ip];
+
+        double pl_int_pt = 0.;
+        double totalrho_int_pt =
+            0.;  // total mass density of the light component
+        NumLib::shapeFunctionInterpolate(local_x, sm.N, pl_int_pt,
+                                         totalrho_int_pt);
+        auto const& wp = _integration_method.getWeightedPoint(ip);
+        const double integration_factor =
```

Comment:
You already use integration point data structure, why not put all other constant per-integration-point-values there?